lib/db/backend/leveldb_backend.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package backend
+
+import (
+	"sync"
+
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+const (
+	dbFlushBatch = 4 << MiB
+)
+
+// leveldbBackend implements Backend on top of a leveldb
+type leveldbBackend struct {
+	ldb     *leveldb.DB
+	closeWG sync.WaitGroup
+}
+
+func (b *leveldbBackend) NewReadTransaction() (ReadTransaction, error) {
+	snap, err := b.ldb.GetSnapshot()
+	if err != nil {
+		return nil, wrapLeveldbErr(err)
+	}
+	return leveldbSnapshot{
+		snap: snap,
+		rel:  newReleaser(&b.closeWG),
+	}, nil
+}
+
+func (b *leveldbBackend) NewWriteTransaction() (WriteTransaction, error) {
+	snap, err := b.ldb.GetSnapshot()
+	if err != nil {
+		return nil, wrapLeveldbErr(err)
+	}
+	return &leveldbTransaction{
+		leveldbSnapshot: leveldbSnapshot{
+			snap: snap,
+			rel:  newReleaser(&b.closeWG),
+		},
+		ldb:   b.ldb,
+		batch: new(leveldb.Batch),
+		rel:   newReleaser(&b.closeWG),
+	}, nil
+}
+
+func (b *leveldbBackend) Close() error {
+	b.closeWG.Wait()
+	return wrapLeveldbErr(b.ldb.Close())
+}
+
+func (b *leveldbBackend) Get(key []byte) ([]byte, error) {
+	val, err := b.ldb.Get(key, nil)
+	return val, wrapLeveldbErr(err)
+}
+
+func (b *leveldbBackend) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	return b.ldb.NewIterator(util.BytesPrefix(prefix), nil), nil
+}
+
+func (b *leveldbBackend) NewRangeIterator(first, last []byte) (Iterator, error) {
+	return b.ldb.NewIterator(&util.Range{Start: first, Limit: last}, nil), nil
+}
+
+func (b *leveldbBackend) Put(key, val []byte) error {
+	return wrapLeveldbErr(b.ldb.Put(key, val, nil))
+}
+
+func (b *leveldbBackend) Delete(key []byte) error {
+	return wrapLeveldbErr(b.ldb.Delete(key, nil))
+}
+
+// leveldbSnapshot implements backend.ReadTransaction
+type leveldbSnapshot struct {
+	snap *leveldb.Snapshot
+	rel  *releaser
+}
+
+func (l leveldbSnapshot) Get(key []byte) ([]byte, error) {
+	val, err := l.snap.Get(key, nil)
+	return val, wrapLeveldbErr(err)
+}
+
+func (l leveldbSnapshot) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	return l.snap.NewIterator(util.BytesPrefix(prefix), nil), nil
+}
+
+func (l leveldbSnapshot) NewRangeIterator(first, last []byte) (Iterator, error) {
+	return l.snap.NewIterator(&util.Range{Start: first, Limit: last}, nil), nil
+}
+
+func (l leveldbSnapshot) Release() {
+	l.snap.Release()
+	l.rel.Release()
+}
+
+// leveldbTransaction implements backend.WriteTransaction using a batch (not
+// an actual leveldb transaction)
+type leveldbTransaction struct {
+	leveldbSnapshot
+	ldb   *leveldb.DB
+	batch *leveldb.Batch
+	rel   *releaser
+}
+
+func (t *leveldbTransaction) Delete(key []byte) error {
+	t.batch.Delete(key)
+	return wrapLeveldbErr(t.checkFlush())
```

Comment:
That reintroduces the problem with non-atomic db updates: Previously `checkFlush` was only called when all db operations that need to go through (or not) together were added to the batch. The only options I see are to either expose `checkFlush`, guaranteeing that the db is only written in those cases, or add a batch-like part to the interface.

Reply:
This is somewhat of an issue. Frankly one of the motivations behind this work is to move to something that provides real transactions (whether that's still goleveldb or something else) and I'm not convinces the current guarantees are what you think. Having multiple writes go out as a batch doesn't make it atomic, it just means there are multiple non-atomic writes happening in close temporal succession... I agree though that this change makes it strictly less atomic, whatever the state before.

For what it's worth the Badger interface is the same. Badger can at any point in a transaction return ErrTransactionTooBig at which point the thing to do is to commit it, create a new one, and continue onwards...

Reply:
Ok, maybe I shouldn't have brought up "atomic". However this clearly makes it worse, as before we were vulnerable to an inconsistent db while "there are multiple non-atomic writes happening in close temporal succession", now we are vulnerable during the entire time an implementation thinks it doesn't yet need to flush a transaction.

> We don't create large transactions though, except in certain "internal" loops like when deleting a folder, as all other updates from the outside come in as small batches anyway.

Possibly, but that means we depend on implementation details, namely that model-side batching creates batches small enough that the backend implementation commits it as a whole.

And I don't know what the properties of a "real transactions" are - could you quickly explain what exactly you want there?

Reply:
What I want is for things to be atomically committed once we call commit, all or nothing. Currently there's no guarantee -- in fact, when there *is* a problem it would probably be quite unlikely for it to happen precisely *between* two batches. I mean, if there's an I/O error or the disk becomes full it will probably happen when we are in fact writing, like in the middle of a batch, not when we are *not* writing. :)

But I'm still thinking of the best way to handle this. We could add some sort of Checkpoint() method which is advisory that if a transaction / batch needs to be broken, this would be a good time to do it... (I.e., that current checkFlush)

Reply:
> What I want is for things to be atomically committed once we call commit, all or nothing. [...]

Cool, I very much want that too - that was my only theory, but I dismissed it as too unrealistic :D 

Well, OOM, power-out, ... are still possibilities.

And how to do it: I was thinking about adding a simple `*Batch` type that contains keys to put/delete and a `Apply(b *Batch)` to the writer interface with the requirement that the writer must flush either before or after all the operations in batch.

Reply:
I don't want more types, as Batch on top of Transaction on top of DB doesn't make sense on its own, imho. But I added the Checkpoint thing to be roughly equivalent with the already existing checkFlush. The behavior is sort of unspecified ("maybe it commits, maybe it doesn't, it might affect your reads") but that's what we already had and what we get from leveldb...

Reply:
With leveldb I don't understand why the read ambiguity is needed. A leveldb snapshot is just the state of a the db at a specific point in time - we can't modify or don't release/resnapshot when flushing the buffer associated with the snapshot making up a read-write-transaction. The only unspecified behaviour I get is that we can't require "checkpointing" to only happen on calling `Checkpoint`, but just tell the implementation should checkpoint then if `Checkpoint` is called regularly. If we do require this ambiguity about future reads on `Checkpoint`, if anyway possible `Checkpoint` should get an additional return value that signifies whether the transaction has been broken thus (further reads can change) or not. 

Another thing that should be stated explicitly: I assume `Commit` is final (i.e. no more writes can/should happen after calling it)?

Reply:
You're right, I was mixing up things. There is a snapshot so there is no ambiguity. There is also no need for such an ambiguity in the future either as we can always have a read-snap + write-tran.